### PR TITLE
[HWORKS-914] Hive services containerization

### PR DESCRIPTION
**Pull request by Jacarte**

### CODE REVIEW

**Comment by SirOibaf** [comment]
Why including the `hive-site.xml` in the container? That should be part of the helm chart.

**Comment by Jacarte** [comment]
> Why including the `hive-site.xml` in the container? That should be part of the helm chart.

Yeap, it is mostly for dev, ideally one should be able of shuting down the vm service and replace it by a docker run. I will remove it, users should provide their hive-site.xml :)

**Comment by Jacarte** [comment]
It is fixed now @SirOibaf

File: dockerfiles/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+# /usr/local/openjdk-8/bin/java
+ENV JAVA_HOME=/usr/local/openjdk-8
+RUN mkdir -p /usr/lib/jvm &&  ln -s $JAVA_HOME /usr/lib/jvm/java-8-openjdk-amd64
+ENV HADOOP_HOME=/srv/hops/hadoop
+ENV HADOOP_CONF_DIR="$HADOOP_HOME/conf"
+
+# Download hive
+# https://archiva.hops.works/repository/Hops/io/hops/hive/hive-packaging/3.0.0.13.3/hive-packaging-3.0.0.13.3-bin.tar.gz
+RUN wget https://archiva.hops.works/repository/Hops/io/hops/hive/hive-packaging/$HIVE_VERSION/hive-packaging-$HIVE_VERSION-bin.tar.gz && \
```
Comment (maismail) [inline dockerfiles/Dockerfile:38]: btw, as far as i know we also have hive-deploy job which builds and copy the version to repo.hops.works

File: dockerfiles/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    ln -s /srv/hops/apache-hive-$HIVE_VERSION-bin /srv/hops/hive && \
+    rm hive-packaging-$HIVE_VERSION-bin.tar.gz
+
+
+ENV HIVE_HOME=/srv/hops/hive
+ENV PATH="/srv/hops/super_crypto/hive:$HIVE_HOME/bin:/opt/venv/bin:$HADOOP_HOME/bin:$PATH"
+
+
+# Download extra jars
+# FIXME: this should come from our own repo
```
Comment (maismail) [inline dockerfiles/Dockerfile:49]: You can use this url instead 
https://repo.hops.works/master/prometheus/jmx_prometheus_javaagent-0.12.0.jar

File: dockerfiles/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ENV HOPSWORKS_JARS=/srv/hops/hopsworks-jars
+RUN mkdir -p ${HOPSWORKS_JARS}
+ARG PROMETHEUS_EXPORTER_VERSION="0.12.0"
+RUN wget -O ${HOPSWORKS_JARS}/jmx_prometheus_javaagent-${PROMETHEUS_EXPORTER_VERSION}.jar  https://repo.maven.apache.org/maven2/io/prometheus/jmx/jmx_prometheus_javaagent/${PROMETHEUS_EXPORTER_VERSION}/jmx_prometheus_javaagent-${PROMETHEUS_EXPORTER_VERSION}.jar
+
+ARG MYSQL_CONNECTOR_VERSION="8.0.21"
+RUN wget -O ${HOPSWORKS_JARS}/mysql-connector-java-${MYSQL_CONNECTOR_VERSION}-bin.jar  https://repo1.maven.org/maven2/mysql/mysql-connector-java/8.0.21/mysql-connector-java-8.0.21.jar
```
Comment (maismail) [inline dockerfiles/Dockerfile:56]: we also have it on repo.hops.works
https://repo.hops.works/master/mysql-connector-java-8.0.21-bin.jar

**Comment by maismail** [comment]
@Jacarte just keeping it here for the future once Antonios is done with the first version of the Jenkins building images, that we should add a Jenkins file here to build the images

File: dockerfiles/Dockerfile (hunk 2 above)
Comment (Jacarte) [reply dockerfiles/Dockerfile]: Fixed

File: dockerfiles/Dockerfile (hunk 3 above)
Comment (Jacarte) [reply dockerfiles/Dockerfile]: Fixed

File: dockerfiles/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+
+ARG HUDI_BUNDLE_VERSION="0.12.3.0"
```
[the remark this answers is not in the document]
Comment (Jacarte) [reply dockerfiles/Dockerfile]: Fixed

File: dockerfiles/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ARG MYSQL_CONNECTOR_VERSION="8.0.21"
+ARG HUDI_BUNDLE_VERSION="0.12.3.0"
+
+# This is for local dev in the VM
+RUN groupadd -g 1505 $GROUP && \
+    useradd -m --uid 1515 --gid 1505 $USER
+RUN groupadd -g 1511 metastore && \
+    usermod -a -G metastore $USER
+RUN groupadd -g 1234 hadoop && \
+    usermod -a -G hadoop $USER
```
Comment (maismail) [inline dockerfiles/Dockerfile:17]: can you add the user and group ids as variables above as well

Comment (maismail) [reply dockerfiles/Dockerfile:17]: also, i'm not sure what is this metastore group?

**Review by maismail — CHANGES_REQUESTED** [review]

File: dockerfiles/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,83 @@
+FROM openjdk:8
+ARG HIVE_VERSION=3.0.0.13.3
+ARG HADOOP_VERSION_EE=3.2.0.12-EE-SNAPSHOT
+ARG PROMETHEUS_EXPORTER_VERSION="0.12.0"
+ARG USER="hive"
+ARG USERID="1515"
```
Comment (maismail) [inline dockerfiles/Dockerfile:6]: from looking at the chef cookbooks, this is not the user_id for hive - it should be `1516` 
that seems to be the user for certs 
https://github.com/logicalclocks/kagent-chef/blob/master/attributes/default.rb#L10C1-L13C60

File: dockerfiles/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ARG GROUP="hive"
```
Comment (maismail) [inline dockerfiles/Dockerfile:7]: that is secure group id `metaserver`
https://github.com/logicalclocks/hops-hadoop-chef/blob/master/attributes/default.rb#L16

File: dockerfiles/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+ARG GROUPID="1505"
+ARG HADOOP_GROUP_ID="1234"
+ARG METASTORE_ID="1511"
```
Comment (maismail) [inline dockerfiles/Dockerfile:10]: that is the certs user id 
https://github.com/logicalclocks/kagent-chef/blob/master/attributes/default.rb#L10C1-L13C60

File: dockerfiles/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+# ENV HIVESERVER2_HADOOP_OPTS='-Djava.io.tmpdir=/tmp/$HIVE_USER'
+
+
+RUN mkdir -p /home/$USER
+RUN chown $USER:$GROUP /home/$USER
+
+WORKDIR /home/$USER
+COPY dockerfiles/entrypoint.sh entrypoint.sh
+RUN chown $USER:$GROUP  entrypoint.sh 
+RUN chown -R $USER:$GROUP $HIVE_HOME
```
Comment (maismail) [inline dockerfiles/Dockerfile:80]: I think everything should be user hive, and group hadoop 
```
root@ip-172-16-4-136:/srv/hops/apache-hive# ls -la
total 236
drwxrwx--- 12 hive hadoop   4096 Oct 13 14:45 .
drwxr-xr-x 39 root hadoop   4096 Oct 13 14:54 ..
-rw-r--r--  1 hive hadoop      0 Oct 13 14:02 .hive_extracted_3.0.0.13.2
-rw-r--r--  1 hive hadoop  20798 Sep 10 15:53 LICENSE
-rw-r--r--  1 hive hadoop    230 Sep 10 15:53 NOTICE
-rw-r--r--  1 hive hadoop 143769 Sep 10 15:53 RELEASE_NOTES.txt
drwxr-xr-x  3 hive hadoop   4096 Oct 30 22:44 bin
drwxr-xr-x  2 hive hadoop   4096 Oct 13 14:02 binary-package-licenses
drwxr-xr-x  2 hive hadoop   4096 Feb  1 18:32 conf
drwxr-x---  2 hive hadoop   4096 Oct 13 14:47 consul
drwxr-xr-x  4 hive hadoop   4096 Oct 13 14:02 examples
drwxr-xr-x  7 hive hadoop   4096 Oct 13 14:02 hcatalog
drwxr-xr-x  2 hive hadoop   4096 Oct 13 14:02 hopsworks-jars
drwxr-xr-x  2 hive hadoop   4096 Oct 13 14:02 jdbc
drwxr-xr-x  4 hive hadoop  20480 Oct 13 14:02 lib
lrwxrwxrwx  1 hive hadoop     36 Oct 13 14:45 logs -> /srv/hopsworks-data/apache-hive/logs
drwxr-xr-x  4 hive hadoop   4096 Oct 13 14:02 scripts
root@ip-172-16-4-136:/srv/hops/apache-hive# su hive
hive@ip-172-16-4-136:/srv/hops/apache-hive$ id
uid=1516(hive) gid=1234(hadoop) groups=1234(hadoop),1505(metaserver)
hive@ip-172-16-4-136:/srv/hops/apache-hive$ groups
hadoop metaserver
hive@ip-172-16-4-136:/srv/hops/apache-hive$
```

**Review by maismail — CHANGES_REQUESTED** [review]

File: dockerfiles/Dockerfile (hunk 9 above)
Comment (Jacarte) [reply dockerfiles/Dockerfile]: Fixed

File: dockerfiles/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+ARG PROMETHEUS_EXPORTER_VERSION="0.12.0"
+ARG MYSQL_CONNECTOR_VERSION="8.0.21"
+ARG HUDI_BUNDLE_VERSION="0.12.3.0"
+
+# This is for local dev in the VM
+RUN useradd -m --uid $USERID  $USER
+
+RUN groupadd -g ${METASERVER_ID} metaserver && \
+    usermod -a -G metaserver $USER
+RUN groupadd -g ${HADOOP_GROUP_ID} hadoop && \
```
Comment (maismail) [inline dockerfiles/Dockerfile:18]: you can add hadoop as an argument as well

File: dockerfiles/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+ENV HIVE_AUX_JARS_PATH=$HOPSWORKS_JARS
+# ENV HIVESERVER2_HADOOP_OPTS='-Djava.io.tmpdir=/tmp/$HIVE_USER'
+
+
+RUN mkdir -p /home/$USER
+RUN chown $USER /home/$USER
+
+WORKDIR /home/$USER
+COPY dockerfiles/entrypoint.sh entrypoint.sh
+RUN chown $USER  entrypoint.sh
```
Comment (maismail) [inline dockerfiles/Dockerfile:77]: I think if you define the hadoop group above you can do the following:

$USER:$HADOOP_GROUP

File: dockerfiles/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+# ENV HIVESERVER2_HADOOP_OPTS='-Djava.io.tmpdir=/tmp/$HIVE_USER'
+
+
+RUN mkdir -p /home/$USER
+RUN chown $USER /home/$USER
+
+WORKDIR /home/$USER
+COPY dockerfiles/entrypoint.sh entrypoint.sh
+RUN chown $USER  entrypoint.sh 
+RUN chown -R $USER $HIVE_HOME
```
Comment (maismail) [inline dockerfiles/Dockerfile:78]: I think if you define the hadoop group above you can do the following:

$USER:$HADOOP_GROUP

File: dockerfiles/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,81 @@
+FROM openjdk:8
```
Comment (maismail) [inline dockerfiles/Dockerfile:1]: how big is the difference between jdk and jre base image i mean if they have any?

**Review by maismail — CHANGES_REQUESTED** [review]

File: dockerfiles/Dockerfile (hunk 12 above)
Comment (Jacarte) [reply dockerfiles/Dockerfile]: Fixed

**Review by maismail — APPROVED** [review]

**Pull request merged**